File: app/scripts/lib/tambalBursa.mjs
/**
 * Tambalan ujung dari arsip bursa — dipakai bersama oleh pembangun turunan
 * harian yang membaca arsip harga.
 *
 * ## Kenapa ada
 *
 * Arsip harga memakai kredensial dan bisa berhenti terisi tanpa satu pun
 * galat: yang tertulis adalah bar bertanggal hari ini dengan volume nol.
 * Arsip bursa tidak memakai kredensial apa pun dan tetap terbit, lengkap
 * dengan tutup/tinggi/rendah, volume, nilai, frekuensi, aliran asing, dan
 * jumlah saham untuk 963 emiten.
 *
 * Keduanya boleh disambung karena terukur SAMA, bukan karena kelihatan mirip:
 * 8.976 pasang emiten-hari (150 emiten × 60 hari), median rasio tutup
 * 1,000000; satu-satunya 96 selisih >0,5% semuanya jatuh di hari yang arsip
 * harganya masih bar hantu. 59 hari lain nol selisih.
 *
 * ## Batasnya, dan kenapa ketat
 *
 * Konvensi kedua sumber BERBEDA di masa lalu — arsip harga menyesuaikan
 * riwayat ke aksi korporasi, bursa melaporkan apa adanya saat itu. Menjahit
 * berhari-hari menumpuk selisih yang tak terlihat. Jadi yang disambung hanya
 * hari yang arsip harga belum punya sama sekali, maksimal MAKS_HARI, dan
 * SELALU di memori — berkas arsip tak pernah ditulis ulang, sehingga panen
 * ulang berikutnya menang tanpa perlu membatalkan apa pun.
 *
 * Ekstraksi dari `bangun-harian-papan.mjs` (29 Agu 2026) supaya pembangun
 * lain tak menyalin logikanya — versi yang disalin akan berbeda diam-diam
 * begitu salah satunya diperbaiki.
 */
import { readFileSync, readdirSync, statSync, existsSync } from 'node:fs'
import { gunzipSync } from 'node:zlib'
import { join } from 'node:path'

/** Lebih dari ini = panen ulang sumbernya, bukan dijahit. */
export const MAKS_HARI = 5

const angka = (x) => {
  const v = Number(x)
  return Number.isFinite(v) && v !== 0 ? v : null
}

/**
 * Lembar → rupiah pakai harga rata-rata hari itu (nilai ÷ volume, dua-duanya
 * dari baris yang sama supaya pembilang dan penyebut satu rumah).
 *
 * Diukur atas 5.590 pasang emiten-hari terhadap angka rupiah sungguhan: arah
 * cocok 98,6%, median rasio 0,9995, kumulatif 1,0023, 93% dalam ±10%.
 *
 * BUKAN taksiran yang dulu dibuang karena miring +33% — yang itu level PASAR,
 * dan harga rata-rata pasar didominasi emiten bervolume besar berharga rendah
 * sementara transaksi asing terkonsentrasi di emiten berharga tinggi. Bias
 * komposisi; per emiten tak punya itu.
 */
export function taksirRupiah(lembar, volume, nilai) {
  const v = Number(volume)
  const n = Number(nilai)
  if (!(v > 0) || !(n > 0)) return null
  return (Number(lembar) || 0) * (n / v)
}

/** Bar 6 kolom `[tanggal, buka, tinggi, rendah, tutup, volume]` — format
 *  arsip harga gabungan. Pembukaan dibiarkan null kalau bursa tak
 *  melaporkannya; yang membacanya wajib menjaganya sendiri. */
export function barEnamKolom(r, iso) {
  const tutup = angka(r.Close)
  if (tutup == null) return null
  return [
    iso,
    angka(r.OpenPrice),
    angka(r.High) ?? tutup,
    angka(r.Low) ?? tutup,
    tutup,
    Number(r.Volume) || 0,
  ]
}

/** Bar 17 kolom format arsip harga sumber. Aliran asing DITAKSIR ke rupiah —
 *  slot itu berisi rupiah, sementara bursa melaporkan lembar; mengisinya apa
 *  adanya memberi angka yang berselisih ribuan kali di kolom yang sama. */
export function barTujuhBelasKolom(r, iso) {
  const tutup = angka(r.Close)
  if (tutup == null) return null
  return [
    iso,
    Math.floor(Date.parse(`${iso}T00:00:00+07:00`) / 1000),
    angka(r.OpenPrice),
    angka(r.High) ?? tutup,
    angka(r.Low) ?? tutup,
    tutup,
    Number(r.Volume) || 0,
    Number(r.Value) || 0,
    Number(r.Frequency) || 0,
    taksirRupiah(r.ForeignBuy, r.Volume, r.Value),
    taksirRupiah(r.ForeignSell, r.Volume, r.Value),
    0,
    0,
    Number(r.ListedShares) || 0,
    0,
    0,
    0,
  ]
}

/** Tanggal bar terakhir yang BERISI — bar hantu (volume nol) tak bersuara. */
export function tanggalBerisiTerakhir(bar, iVolume) {
  let i = bar.length - 1
  while (i > 0 && Number(bar[i]?.[iVolume] ?? 0) === 0) i -= 1
  return bar[i]?.[0] ?? null
}

/**
 * Sampai tanggal berapa sebuah direktori arsip harga BENAR-BENAR berisi —
 * modus atas sampel berkas.
 *
 * Untuk pembangun yang tak punya pass pendahulu (IPO membaca hanya emiten
 * yang pernah IPO, jadi ia tak menyapu seluruh arsip lebih dulu). Membaca
 * 60 berkas jauh lebih murah daripada 962, dan modus atas 60 sudah kokoh:
 * yang dicari tanggal yang dimiliki hampir semua emiten, bukan yang langka.
 */
export function tanggalBerisiDiDir(dir, { ruas = 'd', iVolume = 5, nSampel = 60 } = {}) {
  if (!existsSync(dir)) return null
  const c = new Map()
  for (const f of readdirSync(dir).filter((x) => x.endsWith('.json')).slice(0, nSampel)) {
    let j
    try {
      j = JSON.parse(readFileSync(join(dir, f), 'utf8'))
    } catch {
      continue
    }
    const bar = j?.[ruas]
    if (!Array.isArray(bar) || bar.length === 0) continue
    const t = tanggalBerisiTerakhir(bar, iVolume)
    if (t) c.set(t, (c.get(t) ?? 0) + 1)
  }
  return [...c.entries()].sort((a, b) => b[1] - a[1])[0]?.[0] ?? null
}

/**
 * Berkas arsip bursa yang lebih muda daripada `punyaSampai`, siap disisipkan.
 *
 * Bentuk ini dipakai pembangun yang membaca berkas emiten SATU PER SATU
 * (Screener, IPO, Jago Papan): memuat seluruh bar 962 emiten ke memori dulu
 * hanya untuk menambal satu hari akan menelan lebih dari satu gigabita —
 * sementara tambalannya sendiri cuma beberapa ratus kilobita. Jadi yang
 * dimuat tambalannya, dan penyisipannya terjadi saat tiap emiten dibaca.
 *
 * @returns `{ tanggal: string[], perKode: Map<string, bar[]> }`
 */
export function muatTambalanBursa({ dirBursa, punyaSampai, keBar, log = console }) {
  const kosong = { tanggal: [], perKode: new Map() }
  if (!existsSync(dirBursa) || !punyaSampai) return kosong

  const kandidat = []
  for (const th of readdirSync(dirBursa)) {
    if (!/^\d{4}$/.test(th)) continue
    for (const f of readdirSync(join(dirBursa, th))) {
      const m = f.match(/^(\d{4})(\d{2})(\d{2})\.json\.gz$/)
      if (!m) continue
      const iso = `${m[1]}-${m[2]}-${m[3]}`
      if (iso <= punyaSampai) continue
      const jalur = join(dirBursa, th, f)
      if (statSync(jalur).size < 1000) continue
      kandidat.push([iso, jalur])
    }
  }
  if (kandidat.length === 0) return kosong
  kandidat.sort((a, b) => (a[0] < b[0] ? -1 : 1))

  if (kandidat.length > MAKS_HARI) {
    log.warn(
      `arsip harga tertinggal ${kandidat.length} hari dari arsip bursa ` +
        `(${punyaSampai} → ${kandidat.at(-1)[0]}). Melewati batas tambal ` +
        `${MAKS_HARI} hari; sumbernya perlu dipanen ulang, bukan dijahit.`,
    )
    return kosong
  }

  const tanggal = []
  const perKode = new Map()
  for (const [iso, jalur] of kandidat) {
    let rows
    try {
      rows = JSON.parse(gunzipSync(readFileSync(jalur)).toString('utf8'))?.data
    } catch (e) {
      log.warn(`arsip bursa ${iso} tak terbaca: ${e.message}`)
      continue
    }
    if (!Array.isArray(rows) || rows.length === 0) continue
    let n = 0
    for (const r of rows) {
      const baru = keBar(r, iso)
      if (!baru) continue
      const daftar = perKode.get(r.StockCode)
      if (daftar) daftar.push(baru)
      else perKode.set(r.StockCode, [baru])
      n += 1
    }
    if (n > 0) {
      tanggal.push(iso)
      log.log(`tambal ${iso} dari arsip bursa: ${n} emiten`)
    }
  }
  return { tanggal, perKode }
}

/**
 * Sisipkan bar tambahan ke larik bar satu emiten, di tempat.
 *
 * Bar hantu bertanggal sama DITIMPA; bar berisi tak pernah disentuh — arsip
 * yang sudah punya isinya sendiri selalu menang.
 */
export function sisipkanTambalan(bar, tambahan, iVolume) {
  if (!Array.isArray(bar) || !Array.isArray(tambahan)) return bar
  for (const baru of tambahan) {
    const i = bar.findIndex((b) => b[0] === baru[0])
    if (i === -1) bar.push(baru)
    else if (Number(bar[i][iVolume] ?? 0) === 0) bar[i] = baru
  }
  return bar
}

/**
 * Sambung hari yang belum dimiliki arsip harga, DI MEMORI.
 *
 * @param petaBar   Map kode → larik bar (dimodifikasi di tempat)
 * @param opsi.dirBursa   direktori arsip mentah bursa
 * @param opsi.iVolume    indeks kolom volume di bar
 * @param opsi.keBar      pembentuk bar dari satu baris arsip bursa
 * @param opsi.log        penulis pesan (default console)
 * @returns larik tanggal yang benar-benar ditambal
 */
export function tambalDariArsipBursa(petaBar, { dirBursa, iVolume, keBar, log = console }) {
  if (!existsSync(dirBursa)) return []

  // Sampai tanggal berapa arsip harga BENAR-BENAR berisi — modus, bukan satu
  // emiten acak yang bisa kebetulan disuspensi hari itu.
  const suara = new Map()
  for (const bar of petaBar.values()) {
    if (!Array.isArray(bar) || bar.length === 0) continue
    const t = tanggalBerisiTerakhir(bar, iVolume)
    if (t) suara.set(t, (suara.get(t) ?? 0) + 1)
  }
  const punyaSampai = [...suara.entries()].sort((a, b) => b[1] - a[1])[0]?.[0]
  if (!punyaSampai) return []

  const kandidat = []
  for (const th of readdirSync(dirBursa)) {
    if (!/^\d{4}$/.test(th)) continue
    for (const f of readdirSync(join(dirBursa, th))) {
      const m = f.match(/^(\d{4})(\d{2})(\d{2})\.json\.gz$/)
      if (!m) continue
      const iso = `${m[1]}-${m[2]}-${m[3]}`
      if (iso <= punyaSampai) continue
      const jalur = join(dirBursa, th, f)
      // Arsip 0-baris bertanggal muda = "belum terbit", bukan hari libur.
      // Yang kosong tak boleh dianggap sebagai hari bursa.
      if (statSync(jalur).size < 1000) continue
      kandidat.push([iso, jalur])
    }
  }
  if (kandidat.length === 0) return []
  kandidat.sort((a, b) => (a[0] < b[0] ? -1 : 1))

  if (kandidat.length > MAKS_HARI) {
    log.warn(
      `arsip harga tertinggal ${kandidat.length} hari dari arsip bursa ` +
        `(${punyaSampai} → ${kandidat.at(-1)[0]}). Melewati batas tambal ` +
        `${MAKS_HARI} hari; sumbernya perlu dipanen ulang, bukan dijahit.`,
    )
    return []
  }

  const ditambal = []
  for (const [iso, jalur] of kandidat) {
    let rows
    try {
      rows = JSON.parse(gunzipSync(readFileSync(jalur)).toString('utf8'))?.data
    } catch (e) {
      log.warn(`arsip bursa ${iso} tak terbaca: ${e.message}`)
      continue
    }
    if (!Array.isArray(rows) || rows.length === 0) continue

    let n = 0
    for (const r of rows) {
      const bar = petaBar.get(r.StockCode)
      if (!Array.isArray(bar)) continue
      const baru = keBar(r, iso)
      if (!baru) continue
      const i = bar.findIndex((b) => b[0] === iso)
      // Bar hantu (ada tapi volume nol) DITIMPA; bar berisi tak disentuh.
      if (i === -1) bar.push(baru)
      else if (Number(bar[i][iVolume] ?? 0) === 0) bar[i] = baru
      else continue
      n += 1
    }
    if (n > 0) {
      ditambal.push(iso)
      log.log(`tambal ${iso} dari arsip bursa: ${n} emiten`)
    }
  }
  return ditambal
}

File: scripts/tambal_bursa.py
# -*- coding: utf-8 -*-
"""Tambalan ujung dari arsip bursa — sisi Python.

Kembaran `app/scripts/lib/tambalBursa.mjs`; alasan, batas, dan angka
pengukurannya ditulis lengkap di sana dan tidak diulang di sini supaya tak
ada dua versi yang berbeda diam-diam. Ringkasnya:

* Arsip harga memakai kredensial dan bisa berhenti terisi tanpa satu pun galat
  — yang tertulis bar bertanggal hari ini dengan volume nol.
* Arsip bursa tidak memakai kredensial dan tetap terbit, 963 emiten per hari.
* Keduanya terukur SAMA pada hari yang dua-duanya punya: median rasio tutup
  1,000000 atas 8.976 pasang emiten-hari.
* Yang disambung HANYA hari yang arsip harga belum punya, maksimal MAKS_HARI,
  dan selalu di memori — berkas arsip tak pernah ditulis ulang.

Dipakai `scripts/riset/kartu_analisa.py`. Bentuknya sengaja "muat sekali,
sisipkan per emiten": kartu membaca 963 berkas satu per satu, dan memuat
semuanya ke memori dulu hanya untuk menambal satu hari akan menelan lebih
dari satu gigabita.
"""
from __future__ import annotations

import gzip
import json
import re
from pathlib import Path

AKAR = Path(__file__).resolve().parent.parent
DIR_BURSA = AKAR / "_arsip-mentah" / "asing"

#: Lebih dari ini = panen ulang sumbernya, bukan dijahit.
MAKS_HARI = 5

_NAMA = re.compile(r"^(\d{4})(\d{2})(\d{2})\.json\.gz$")


def _angka(x) -> float | None:
    try:
        v = float(x)
    except (TypeError, ValueError):
        return None
    return v if v else None


def bar_enam_kolom(r: dict, iso: str) -> list | None:
    """`[tanggal, buka, tinggi, rendah, tutup, volume]` — format `ohlc/`.

    Pembukaan dibiarkan None kalau bursa tak melaporkannya (terukur 28 Agu
    2026: kosong di 220 dari 833 emiten aktif). Yang membacanya wajib
    menjaganya sendiri; mengisinya nol memberi angka yang terbaca seperti
    hasil hitungan sungguhan.
    """
    tutup = _angka(r.get("Close"))
    if tutup is None:
        return None
    return [
        iso,
        _angka(r.get("OpenPrice")),
        _angka(r.get("High")) or tutup,
        _angka(r.get("Low")) or tutup,
        tutup,
        float(r.get("Volume") or 0),
    ]


def tanggal_berisi_terakhir(baris: list, i_volume: int = 5) -> str | None:
    """Tanggal bar terakhir yang BERISI — bar hantu tak bersuara."""
    if not baris:
        return None
    i = len(baris) - 1
    while i > 0 and not (baris[i][i_volume] if len(baris[i]) > i_volume else 0):
        i -= 1
    return baris[i][0] if baris[i] else None


def tanggal_berisi_di_dir(dir_ohlc: Path, ruas: str = "d", i_volume: int = 5,
                          n_sampel: int = 60) -> str | None:
    """Sampai tanggal berapa direktori arsip harga benar-benar berisi (modus).

    Modus atas 60 berkas, bukan atas semuanya: yang dicari tanggal yang
    dimiliki hampir semua emiten, bukan yang langka — dan 60 sudah kokoh
    untuk itu, sementara 963 berkas mahal dibaca dua kali.
    """
    if not dir_ohlc.exists():
        return None
    hitung: dict[str, int] = {}
    for p in sorted(dir_ohlc.glob("*.json"))[:n_sampel]:
        try:
            d = json.loads(p.read_text(encoding="utf-8"))
        except Exception:  # noqa: BLE001
            continue
        baris = d.get(ruas) if isinstance(d, dict) else None
        if not baris:
            continue
        t = tanggal_berisi_terakhir(baris, i_volume)
        if t:
            hitung[t] = hitung.get(t, 0) + 1
    return max(hitung.items(), key=lambda kv: kv[1])[0] if hitung else None


def muat_tambalan(punya_sampai: str | None, ke_bar=bar_enam_kolom,
                  lapor=print) -> tuple[list[str], dict[str, list]]:
    """Bar arsip bursa yang lebih muda daripada `punya_sampai`, per emiten.

    Mengembalikan `(tanggal_ditambal, {kode: [bar, ...]})`.
    """
    if not punya_sampai or not DIR_BURSA.exists():
        return [], {}

    kandidat: list[tuple[str, Path]] = []
    for th in DIR_BURSA.iterdir():
        if not th.is_dir() or not re.fullmatch(r"\d{4}", th.name):
            continue
        for f in th.iterdir():
            m = _NAMA.match(f.name)
            if not m:
                continue
            iso = f"{m[1]}-{m[2]}-{m[3]}"
            if iso <= punya_sampai:
                continue
            # Arsip 0-baris bertanggal muda = "belum terbit", bukan hari libur.
            if f.stat().st_size < 1000:
                continue
            kandidat.append((iso, f))
    if not kandidat:
        return [], {}
    kandidat.sort()

    if len(kandidat) > MAKS_HARI:
        lapor(f"  arsip harga tertinggal {len(kandidat)} hari dari arsip bursa "
              f"({punya_sampai} -> {kandidat[-1][0]}); melewati batas tambal "
              f"{MAKS_HARI} hari, sumbernya perlu dipanen ulang bukan dijahit")
        return [], {}

    tanggal: list[str] = []
    per_kode: dict[str, list] = {}
    for iso, jalur in kandidat:
        try:
            with gzip.open(jalur, "rt", encoding="utf-8") as fh:
                rows = json.load(fh).get("data")
        except Exception as e:  # noqa: BLE001
            lapor(f"  arsip bursa {iso} tak terbaca: {e}")
            continue
        if not rows:
            continue
        n = 0
        for r in rows:
            baru = ke_bar(r, iso)
            if baru is None:
                continue
            per_kode.setdefault(r.get("StockCode"), []).append(baru)
            n += 1
        if n:
            tanggal.append(iso)
            lapor(f"  tambal {iso} dari arsip bursa: {n} emiten")
    return tanggal, per_kode


def sisipkan(baris: list, tambahan: list | None, i_volume: int = 5) -> list:
    """Sisipkan bar tambahan ke deret satu emiten, di tempat.

    Bar hantu bertanggal sama DITIMPA; bar berisi tak pernah disentuh — arsip
    yang sudah punya isinya sendiri selalu menang, jadi panen ulang berikutnya
    otomatis mengembalikan angka aslinya tanpa perlu membatalkan apa pun.
    """
    if not tambahan:
        return baris
    indeks = {r[0]: i for i, r in enumerate(baris)}
    for baru in tambahan:
        i = indeks.get(baru[0])
        if i is None:
            baris.append(baru)
        elif not (baris[i][i_volume] if len(baris[i]) > i_volume else 0):
            baris[i] = baru
    return baris


def _uji() -> None:
    """Swauji tanpa menyentuh cakram."""
    d = [["2026-08-26", 1, 1, 1, 100, 500], ["2026-08-27", 1, 1, 1, 110, 0]]
    assert tanggal_berisi_terakhir(d) == "2026-08-26", "bar hantu tak boleh menang"

    # Bar hantu ditimpa, bar berisi tidak.
    baris = [["2026-08-26", 1, 1, 1, 100, 500], ["2026-08-27", 1, 1, 1, 110, 0]]
    sisipkan(baris, [["2026-08-27", 2, 2, 2, 120, 900], ["2026-08-28", 3, 3, 3, 130, 700]])
    assert baris[1][4] == 120 and baris[1][5] == 900, "bar hantu wajib ditimpa"
    assert baris[0][4] == 100, "bar berisi tak boleh disentuh"
    assert baris[2][0] == "2026-08-28", "hari baru wajib ditambahkan"

    # Pembukaan kosong tetap None — bukan nol.
    b = bar_enam_kolom({"Close": 100, "High": 0, "Low": 0, "Volume": 5, "OpenPrice": 0}, "2026-08-28")
    assert b[1] is None, "pembukaan kosong wajib None, bukan 0"
    assert b[2] == 100 and b[3] == 100, "tinggi/rendah kosong jatuh ke tutup"

    # Tanpa harga tutup tak ada yang bisa dihitung.
    assert bar_enam_kolom({"Close": 0, "Volume": 5}, "2026-08-28") is None

    print("uji tambal_bursa: LOLOS")


if __name__ == "__main__":
    _uji()
